cumulusci/utils/clean_invalid_references.py

```diff
@@ -0,0 +1,365 @@
+            root = ET.parse(file).getroot()
+            root = strip_namespace(root)
+            for key, value in PACKAGEXML_DICT.items():
+                if isinstance(value, dict):
+                    for perm, parent in value.items():
+                        package_xml_input.setdefault(key, set()).update(
+                            fetch_permissionable_entity_names(root, perm, parent)
+                        )
+                else:
+                    package_xml_input.setdefault(key, set()).update(
+                        fetch_permissionable_entity_names(root, value)
+                    )
+    # Remove any entities with no entries
+    package_xml_input = {
+        key: value for key, value in package_xml_input.items() if value
+    }
+    package_xml = create_package_xml(
+        input_dict=package_xml_input, api_version=api_version
+    )
+    return package_xml
+
+
+def create_package_xml(input_dict: dict, api_version: str):
+    package_xml = '<?xml version="1.0" encoding="UTF-8"?>\n'
+    package_xml += '<Package xmlns="http://soap.sforce.com/2006/04/metadata">\n'
+
+    for name, members in input_dict.items():
+        package_xml += "    <types>\n"
+        for member in members:
+            package_xml += f"        <members>{member}</members>\n"
+        package_xml += f"        <name>{name}</name>\n"
+        package_xml += "    </types>\n"
+
+    package_xml += f"    <version>{api_version}</version>\n"
+    package_xml += "</Package>\n"
+    return package_xml
+
+
+def get_fields_and_recordtypes(root: ET.Element, objectName):
+    fields = set()
+    recordTypes = set()
+
+    for field in root.findall("fields"):
+        fields.update([objectName + "." + field.find("fullName").text])
+
+    for recordType in root.findall("recordTypes"):
+        recordTypes.update([objectName + "." + recordType.find("fullName").text])
+
+    return fields, recordTypes
+
+
+def get_tabs_from_app(root: ET.Element):
+    tabs = set()
+    for tab in root.findall("tabs"):
+        tabs.update([tab.text])
+    return tabs
+
+
+def get_target_entities_from_zip(zip_src: zipfile.ZipFile):
+    target_entities = {key: set() for key in FOLDER_PERM_DICT.keys()}
+    for name in zip_src.namelist():
+        if name == "package.xml":
+            continue
+        metadataType = name.split("/")[0]
+        metadataName = name.split("/")[1].split(".")[0]
+        target_entities[metadataType].update([metadataName])
+
+        # If object, fetch all the fields and record types present inside the file
+        if name.endswith(".object"):
+            file = zip_src.open(name)
+            root = ET.parse(file).getroot()
+            root = strip_namespace(root)
+            fields, recordTypes = get_fields_and_recordtypes(root, metadataName)
+            target_entities["fields"].update(fields)
+            target_entities["recordTypes"].update(recordTypes)
+
+        # To handle tabs which are not part of TabDefinition Table
```

Comment:
This is a weird new facet I had not seen before. So basically you find situations where a CustomApplication contains tabs that are not found in TabDefinition?

Reply:
Yes there are tabs which are present in CustomApplications which are not inside TabDefinition. But I changed to using `PermissionSetTabSetting` which covers everything. This block of code is not needed